In [ ]:
import numpy as np
import sympy as sp

## Практическое задание **VII.9.5 пункт в)**

Для функции, заданной таблично, вычислить значение определенного интеграла методом трапеций, сделать уточнение результата экстраполяцией Ричардсона. Сравнить уточненный результат с вычислениями по методу Симпсона.

|      $x$       |      $f(x)$     |
| :------------: | :-------------: |
|    $x_1 = 0$   |   $1.000000$    |
|  $x_2 = 0.15$  |   $1.007568$    |
|  $x_3 = 0.3$   |   $1.031121$    |
|  $x_4 = 0.45$  |   $1.073456$    |
|  $x_5 = 0.6$   |   $1.140228$    |
|  $x_6 = 0.75$  |   $1.242129$    |
|  $x_7 = 0.9$   |   $1.400176$    |
|  $x_8 = 1.05$  |   $1.660300$    |
|  $x_9 = 1.2$   |   $2.143460$    |


Заведем массив точек. В этой задаче не буду создавать классы в виду нецелесообразности.

In [ ]:
points = [
          (0, 1),
          (0.15, 1.007568),
          (0.3, 1.031121),
          (0.45, 1.073456),
          (0.6, 1.140228),
          (0.75, 1.242129),
          (0.9, 1.400176),
          (1.05, 1.660300),
          (1.2, 2.143460)
]

Создадим метод для обсчёта интеграла методом трапеций. Так же в метод добавим возможность указывать шаг сетки. Соответственно для поиска $I_{2h}$ необходимо будет вызывать метод с параметром `step = 2`.

In [ ]:
def trapezoidal(points, step = 1):
  h = (points[-1][0] - points[0][0]) / ((len(points) - 1) / step)

  considered = [points[i] for i in range(0, len(points), step)]

  integral = h * ((considered[0][1] + considered[-1][1]) / 2 + 
                  sum(list(map(lambda x: x[1], considered))[1:len(considered) - 1]))

  return integral

Теперь подсчитаем с помощью созданного метода $I_h$ и $I_{2h}$

In [ ]:
print(f"Iₕ = {trapezoidal(points)}")
print(f"I₂ₕ = {trapezoidal(points, 2)}")

Iₕ = 1.5190062
I₂ₕ = 1.5429765


В результате мы получили результат 

$$
  I_h = 1.5190062
  \hspace{1cm}
  I_{2h} = 1.5429765
$$

Полученный резултат совпадает с верным ответом из задачника.

Теперь напишем метод для уточнения результатов, полученных методом трапеций. Считать $I_h$ и $I_{nh}$ будем внутри метода, чтобы попутно рассчитать $h$.

In [ ]:
def richardsonExtrapolation(points, h_power, next_grid = 2):
  Ih = trapezoidal(points, step = 1)
  Inh = trapezoidal(points, step = next_grid)

  h = (points[-1][0] - points[0][0]) / (len(points) - 1)

  chp = (Ih - Inh) / (next_grid ** h_power - 1)

  return Ih + chp

Напишем так же метод для рассчета интеграла методом Симпсона.

In [ ]:
def simpson(points):
  result = 0

  n = len(points) - 1
  h = (points[-1][0] - points[0][0]) / n

  middle4 = sum(list(map(lambda x: 4 * x[1], points[1:n:2])))
  middle2 = sum(list(map(lambda x: 2 * x[1], points[2:n:2])))
  
  return (h / 3) * (points[0][1] + points[-1][1] + middle4 + middle2)

Рассчитаем значение интеграла методом трапеций с уточнением и методом Симпсона.

In [ ]:
print(f"Richardson: I = {richardsonExtrapolation(points, 2, 2)}")
print(f"Simson:     I = {simpson(points)}")

Richardson: I = 1.5110161
Simson:     I = 1.5110161


$$
  I_R = 1.5110161
  \hspace{1cm}
  I_S = 1.5110161
$$

Сразу видно, что значения, полученные данными методами полностью совпадают.